x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java

```diff
@@ -548,61 +548,100 @@ public void checkPrivileges(
                 listener.onResponse(result);
                 return;
             }
+            maybeCachingListener = listener.delegateFailure((delegateListener, privilegesCheckResult) -> {
+                try {
+                    simpleRole.cacheHasPrivileges(settings, privilegesToCheck, privilegesCheckResult);
+                } catch (ExecutionException e) {
+                    logger.error("Failed to cache check result for [{}]", privilegesToCheck);
+                    delegateListener.onFailure(e);
+                    return;
+                }
+                delegateListener.onResponse(privilegesCheckResult);
+            });
+        } else {
+            // caching of check result unsupported
+            maybeCachingListener = listener;
         }
 
-        Map<String, Boolean> cluster = new HashMap<>();
+        boolean allMatch = true;
+
+        final Map<String, Boolean> clusterPrivilegesCheckResults = new HashMap<>();
         for (String checkAction : privilegesToCheck.cluster()) {
-            cluster.put(checkAction, userRole.grants(ClusterPrivilegeResolver.resolve(checkAction)));
+            boolean privilegeGranted = userRole.grants(ClusterPrivilegeResolver.resolve(checkAction));
+            allMatch = allMatch && privilegeGranted;
+            if (privilegesToCheck.runDetailedCheck()) {
+                clusterPrivilegesCheckResults.put(checkAction, privilegeGranted);
+            } else if (false == privilegeGranted) {
```

Comment:
Nit: I feel it is more natural to test `allMatch` instead. It is after all the actual value we want to return (details or not).

Reply:
Okay, I made it so.